### ISSUE TITLE
[BUG][Fuzzer][wrong-code] `T.im2col` (Hopper TMA) silently builds a wrong im2col matrix (no error) when `OW != W` — window origins overrun the input row instead of wrapping to the next one

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2426]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.11

### System information

- Installed from source (commit `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`, tag `v0.1.11`).
- Python 3.13, Linux.
- GPU: NVIDIA H100 80GB HBM3 (sm_90). Hopper-specific: the defect is in the TMA im2col path (`Im2Col::Lower`, emitting `cp.async.bulk.tensor.4d…im2col`), selected only on sm_90.

```python
import sys, tilelang, torch
print(sys.version.split()[0], sys.platform)                              # 3.13.13 linux
print(tilelang.__version__)                                              # 0.1.11
print(torch.__version__, torch.version.cuda)                             # 2.11.0+cu130 13.0
print(torch.cuda.get_device_name(), torch.cuda.get_device_capability())  # NVIDIA H100 80GB HBM3 (9, 0)
```

### Problem description

`T.im2col` on Hopper (sm_90) silently builds a wrong im2col matrix whenever the output width differs from the input width (`OW != W`): the matrix's row index advances on the **output** grid (a row should start a new output pixel, wrapping every `OW`), but the TMA load it lowers to advances each tile's window origin on the **input** grid (wrapping every `W`). So once a tile spans more than one output row, every row past the first reads a window from the wrong input position — and the window origin is pushed onto columns `OW .. W-1`, so the window itself overruns the input-row edge. No `T.gemm` or convolution is needed to see it: the gathered matrix is already wrong, and the kernel compiles and runs with no error.

With exact-integer fp16 inputs (so the reference is bit-exact), `N=1, C=32, H=W=16, K=3, S=1, P=0` (→ `OH=OW=14 != W=16`), `block_M=64`, comparing the gathered matrix directly against `torch.nn.functional.unfold`: **36918 / 56448** elements wrong, deterministic. The signature is that rows `0 .. OW-1` (output row `oh=0`) are correct and the first wrong row is exactly row `OW` (the start of `oh=1`).

Concretely, row `OW=14` (which should hold the window for output pixel `oh=1, ow=0`, i.e. input-row origin `h=1, w=0`) instead reads its window starting at `h=0, w=14`, so the window spans input columns `w = 14, 15, 16` — column `16` is past the width-16 input row and comes back as `0` (encoding each input position as `h*100 + w`):

```
row 14 expected (h*100+w per window cell)   row 14 actual (gathered)
 [100 101 102]   (h=1, w=0,1,2)              [ 14  15   0]   (h=0, w=14,15, then w=16 overruns -> 0)
 [200 201 202]                               [114 115   0]
 [300 301 302]                               [214 215   0]
```

The same program with `OW == W` (e.g. `P=1` → `OW=16`) is bit-exact on the same TMA im2col path, and forcing `block_M` so a tile spans only one output row (`block_M <= OW`) is bit-exact even at `OW != W` — isolating the defect to the window origin advancing on the input grid rather than the output grid (controls below).

### Reproducible example code

(Use a real `.py` file — TileLang's eager front end calls `inspect.getsource`, so `python -c '…'` trips an unrelated `OSError`. Run on an sm_90 GPU so the TMA im2col path is selected. The `T.gemm` is dropped: `T.im2col` gathers straight into shared, `T.copy`s out, and the gathered matrix is compared directly to `torch.nn.functional.unfold` — so the wrong gather is exposed by the im2col primitive alone, with no matmul.)

```python
import tilelang
tilelang.disable_cache()
import torch
import tilelang.language as T


@tilelang.jit
def im2col_only(N, C, H, W, K, S, D, P, block_M, block_K, dtype=T.float16):
    OH = (H + 2 * P - D * (K - 1) - 1) // S + 1
    OW = (W + 2 * P - D * (K - 1) - 1) // S + 1
    M = N * OH * OW          # rows of the im2col matrix (one per output pixel)
    KK = K * K * C           # cols of the im2col matrix

    @T.prim_func
    def main(data: T.Tensor((N, H, W, C), dtype),
             col: T.Tensor((M, KK), dtype)):
        with T.Kernel(T.ceildiv(M, block_M), T.ceildiv(KK, block_K)) as (bm, bk):
            sh = T.alloc_shared((block_M, block_K), dtype)
            T.im2col(data, sh, bm, bk, K, S, D, P)
            T.copy(sh, col[bm * block_M, bk * block_K])
    return main


def ref_im2col(data, K, S, D, P):
    x = data.permute(0, 3, 1, 2).float()                  # N,H,W,C -> N,C,H,W
    cols = torch.nn.functional.unfold(x, (K, K), dilation=D, padding=P, stride=S)  # (N, C*K*K, L)
    N, CKK, L = cols.shape
    C = x.shape[1]
    # unfold layout is (C,KH,KW); tilelang im2col col layout is (KH,KW,C). reorder.
    cols = cols.view(N, C, K, K, L).permute(0, 2, 3, 1, 4).reshape(N, K * K * C, L)
    return cols.permute(0, 2, 1).reshape(N * L, K * K * C)  # (N*OH*OW, KK)


def check(label, N, C, H, W, K, S, D, P, block_M, block_K):
    OH = (H + 2 * P - D * (K - 1) - 1) // S + 1
    OW = (W + 2 * P - D * (K - 1) - 1) // S + 1
    M, KK = N * OH * OW, K * K * C
    torch.manual_seed(0)
    data = torch.randint(-2, 3, (N, H, W, C)).cuda().half()
    col = torch.zeros(M, KK, device="cuda", dtype=torch.float16)
    knl = im2col_only(N, C, H, W, K, S, D, P, block_M, block_K)
    knl(data, col)
    ref = ref_im2col(data, K, S, D, P).half()
    o, r = col.float(), ref.float()
    nbad = ((o - r).abs() > 0.5).sum().item()
    maxd = (o - r).abs().max().item()
    tma = "tma_load_im2col" in knl.get_kernel_source()
    bad_rows = [m for m in range(min(M, OW * 3)) if (o[m] - r[m]).abs().max().item() > 0.5]
    print(f"[{label}] OW={OW} W={W} OH={OH} M={M} KK={KK} block_M={block_M} "
          f"tma_im2col={tma} nbad={nbad}/{o.numel()} maxdiff={maxd} "
          f"first_bad_rows={bad_rows[:6]}")


print("=== im2col-only (NO gemm): OW != W -> wrong gather ===")
check("WRONG OW!=W   ", 1, 32, 16, 16, 3, 1, 1, 0, block_M=64, block_K=32)   # OW=14 != W=16
print("=== control: OW == W (P=1) -> exact ===")
check("OK    OW==W   ", 1, 32, 16, 16, 3, 1, 1, 1, block_M=64, block_K=32)   # OW=16 == W=16
print("=== control: block_M=OW -> one output row per tile -> exact even at OW!=W ===")
check("OK    bM=OW   ", 1, 32, 18, 18, 3, 1, 1, 0, block_M=16, block_K=32)   # OW=16, block_M=16
```

Output (deterministic, identical across `num_stages`):

```
=== im2col-only (NO gemm): OW != W -> wrong gather ===
[WRONG OW!=W   ] OW=14 W=16 OH=14 M=196 KK=288 block_M=64 tma_im2col=True nbad=36918/56448 maxdiff=4.0 first_bad_rows=[14, 15, 16, 17, 18, 19]
=== control: OW == W (P=1) -> exact ===
[OK    OW==W   ] OW=16 W=16 OH=16 M=256 KK=288 block_M=64 tma_im2col=True nbad=0/73728 maxdiff=0.0 first_bad_rows=[]
=== control: block_M=OW -> one output row per tile -> exact even at OW!=W ===
[OK    bM=OW   ] OW=16 W=18 OH=16 M=256 KK=288 block_M=16 tma_im2col=True nbad=0/73728 maxdiff=0.0 first_bad_rows=[]
```

The wrong gather is visible in the raw im2col matrix with no matmul: at `OW != W` the first wrong row is `14` (exactly `OW`, i.e. the start of output row `oh=1` — the first row past the tile corner), while the `OW == W` control and the one-output-row-per-tile (`block_M = OW`) control are both bit-exact on the same TMA im2col path. The only varying factor that flips correct↔wrong is whether a tile's load box can cross an output-row boundary.

### Traceback

```pytb
No traceback — the kernel compiles and runs to completion; the result is silently wrong and deterministic.
```

### Expected behavior

`T.im2col` should gather the window for each output pixel `(n, oh, ow)` from its correct input position for any legal geometry, including `OW != W` (any `P=0`, `S>1`, or non-square `H != W`) — i.e. each matrix row should start a new output pixel and wrap to the next input row every `OW` columns, not run the window origin off the input-row edge. The fix should make the gathered matrix match `torch.nn.functional.unfold` for `OW != W`, the way it already does for `OW == W`. TileLang already ships the correct decode for this op: the SIMT path [`LowerIm2ColSIMT`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/op/copy.cc#L148) decodes every element with `out_w = OW` / `out_hw = OH*OW` and handles `OW != W` — so the Hopper TMA lowering is the one place that only respects the output grid at the tile corner.

### Additional context

**Root cause.** The im2col matrix has one row per output pixel (consecutive rows wrap at `OW`). On Hopper, [`Im2Col::Lower`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/cuda/op/copy.cc#L2104) describes the load as a single per-tile **corner** plus a pixel **box**, and decodes that corner correctly on the output grid using `OW`/`OH` — but only [once per `block_M` tile](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/cuda/op/copy.cc#L2197-L2205):

```cpp
// w_dim = OW, h_dim = OH (output dims)
global_coords.push_back(stride * FloorMod(nhw_step * smem_box_pixel, w_dim) - padding);                  // ow
global_coords.push_back(stride * FloorMod(FloorDiv(nhw_step * smem_box_pixel, w_dim), h_dim) - padding); // oh
global_coords.push_back(FloorDiv(nhw_step * smem_box_pixel, w_dim * h_dim));                              // n
```

The pixel-box extent is then set to the destination tile's row count — [`desc.smem_box_pixel = block_M`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/cuda/op/copy.cc#L2144) — a single linear span, and the descriptor's spatial bounds carry only the padding halo ([`lower_corner = upper_corner = {-padding, -padding}`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/cuda/op/copy.cc#L2142)), with the per-load [`offset_w/offset_h`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/cuda/op/copy.cc#L2181) carrying the kernel `(kh, kw)` position. So the emitted [`tma_load_im2col`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/tl_templates/cuda/copy_sm90.h#L377-L409) advances `block_M` pixels linearly along the descriptor's **input-`W`** `global_shape`. `OW` is computed (it sets the corner) but is **not** used to bound that span. So when `block_M` spans more than one output row, the row index wraps at `OW` while the box advances at `W`, and every row past the first drifts by a cumulative `(W - OW) * stride` input columns — the `oh=0`-correct / `oh>=1`-wrong signature, and why a tile confined to one output row is correct. The converged boundary (operator-confirmed, 42/42 configs) is that the gather is correct **iff** `ceil(W/S) == OW` **or** (`block_M <= OW` and `OW % block_M == 0`) — i.e. iff a tile's linear box can never span an output-row boundary; `OW == W` is just the `S=1` instance of `ceil(W/S) == OW`. The shipped [`example_convolution.py`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/examples/convolution/example_convolution.py#L43) masks it because its defaults (`S=1, P=1`) give `OW == W`.

(I'm describing how TileLang configures the descriptor, not asserting a hardware limit. The Hopper im2col instruction exposes corner + halo (`lower_corner`/`upper_corner`) + per-load kernel offsets; whether a single load can be configured to cover a full `OW != W` tile correctly — i.e. whether this is really a descriptor-configuration issue rather than a need to split — is worth confirming against the PTX semantics. Either way the observed defect is that the current `block_M`-as-linear-box configuration gathers wrong for `OW != W`.)

<details>
<summary>Full failing-boundary sweep (42/42 configs, incl. S&gt;1 / D&gt;1 / asymmetric P)</summary>

`S>1`, `D>1`, and asymmetric `P` are additional triggers of the *same* defect, not a separate stride bug: the within-row stride (`elem_stride={1,S,S,1}`, [`copy.cc#L2141`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/cuda/op/copy.cc#L2141)) and dilation offset are exact in isolation; the only failure is the corner-vs-box-wrap mismatch. Forcing `block_M | OW` is bit-exact even at `S=2`/`D=2`; the same config with a multi-output-row tile reverts to wrong on the odd `oh`.

"Correct iff a tile's box can never span an output-row boundary", i.e. `ceil(W/S) == OW` or (`block_M <= OW` and `OW % block_M == 0`):

| config | S | OW | W | ceil(W/S) | block_M | safe? | observed |
|---|---|---|---|---|---|---|---|
| k3s1p1 | 1 | 16 | 16 | 16 | 64 | yes (OW==W) | correct |
| k1s1p0 | 1 | 16 | 16 | 16 | 64 | yes (OW==W) | correct |
| k3s1p0 | 1 | 14 | 16 | 16 | 64 | no | wrong |
| k3s2p0 | 2 | 8 | 17 | 9 | 64 | no | wrong |
| OW32 W34 bM32 | 1 | 32 | 34 | 34 | 32 | yes (bM divides OW) | correct |
| OW32 W34 bM64 | 1 | 32 | 34 | 34 | 64 | no | wrong |
| S2_bM16 | 2 | 16 | 33 | 17 | 16 | yes (bM==OW) | correct |
| D2_bM32 | 1 | 32 | 36 | 36 | 32 | yes (bM divides OW) | correct |
| D2_bM64 | 1 | 32 | 36 | 36 | 64 | no | wrong |
| S2D2_bM16 | 2 | 17 | 37 | 19 | 16 | no (17%16≠0) | wrong |

The `S2D2_bM16` row (`block_M=16 < OW=17` but `17 % 16 ≠ 0` → still wrong) is why the divisibility term, not merely `block_M <= OW`, is needed. Results are identical across `num_stages` (`ns=1` vs `ns=3` bitwise equal), so it is not a pipeline/race effect.

</details>

**Suggested fix.** The current `smem_box_pixel = block_M` treats the box as one linear span on the input grid, which only matches the output-pixel order when `OW == W`. A few options:

1. If the Hopper im2col descriptor can be configured (corner + halo + box) so a single load covers a full `OW != W` tile correctly, that's the cleanest fix — i.e. use `OW` to bound the box, not just to set the corner. (I haven't confirmed the descriptor can express this in one load.)
2. Otherwise, emit one im2col load **per output row** within the tile (recomputing the corner each time), so each load stays within a single output row. The one-output-row-per-tile control shows this is already exact at `OW != W`.
3. Or fall back to the existing SIMT decode (`LowerIm2ColSIMT`, which handles `OW != W` correctly) when a tile could span an output-row boundary.
4. At least reject the unsafe geometry at compile time, so the silent wrong result becomes a clear error.

Option 4 is the smallest change and I confirmed it works (the `OW != W` cases stop compiling, the `OW == W` ones still run correctly).

This issue should be distinct.
